File: sdks/typescript/examples/ask_example.ts
#!/usr/bin/env ts-node
/**
 * Example: Using AvocadoDB's ask() method with TinyLlama
 *
 * This demonstrates the new v2.0 feature for getting natural language
 * answers from your codebase.
 *
 * Requirements:
 *   npm install avocadodb
 *   pip install avocadodb[llm]  # Python SDK with LLM support
 *
 * Setup:
 *   1. Start AvocadoDB server: ./target/release/avocado-server
 *   2. Ingest documents: ./target/release/avocado ingest test-docs/ --recursive
 *   3. Run: ts-node examples/ask_example.ts
 */

import { AvocadoDB } from '../src/client';

async function main() {
  console.log('🥑 AvocadoDB v2.0 - Ask Questions with TinyLlama\n');
  console.log('='.repeat(60));

  // Connect to AvocadoDB server
  const db = new AvocadoDB('http://localhost:8765');

  try {
    const stats = await db.stats();
    console.log('✅ Connected to AvocadoDB');
    console.log(`   Database: ${stats.artifacts} artifacts, ${stats.spans} spans\n`);
  } catch (error) {
    console.error('❌ Cannot connect to AvocadoDB server:', error);
    console.error('   Start server with: ./target/release/avocado-server');
    return;
  }

  // Example queries
  const queries = [
    'How does the compile function work?',
    'What is the WorkingSet structure?',
    'How does span extraction work?',
  ];

  for (const query of queries) {
    console.log(`Question: ${query}`);
    console.log('-'.repeat(60));

    try {
      // Use ask() method - automatically uses TinyLlama if available
      const answer = await db.ask(query, { llm: 'auto' });
      console.log(`Answer: ${answer}\n`);
    } catch (error) {
      console.error(`❌ Error: ${error}\n`);
      // Fallback to just context
      try {
        const context = await db.compile(query);
        console.log(`Context (fallback): ${context.text.substring(0, 200)}...\n`);
      } catch {
        // Ignore
      }
    }
  }

  console.log('='.repeat(60));
  console.log('\n💡 Tips:');
  console.log("   - Use llm: 'auto' (default) to try LLM, fallback to context");
  console.log("   - Use llm: 'local' to require TinyLlama");
  console.log("   - Use llm: 'none' to just get context (same as compile())");
}

if (require.main === module) {
  main().catch(console.error);
}

File: sdks/python/examples/ask_example.py
#!/usr/bin/env python3
"""
Example: Using AvocadoDB's ask() method with TinyLlama

This demonstrates the new v2.0 feature for getting natural language
answers from your codebase.

Requirements:
    pip install avocadodb[llm]  # Includes TinyLlama support
    # Or: pip install avocadodb transformers torch

Setup:
    1. Start AvocadoDB server: ./target/release/avocado-server
    2. Ingest documents: ./target/release/avocado ingest test-docs/ --recursive
    3. Run: python examples/ask_example.py
"""

import sys
from pathlib import Path

# Add parent directory to path
sys.path.insert(0, str(Path(__file__).parent.parent))

from avocado import AvocadoDB


def main():
    print("🥑 AvocadoDB v2.0 - Ask Questions with TinyLlama\n")
    print("=" * 60)
    
    # Connect to AvocadoDB server
    try:
        db = AvocadoDB(url="http://localhost:8765")
        stats = db.stats()
        print("✅ Connected to AvocadoDB")
        print(f"   Database: {stats.get('artifacts_count', 0)} artifacts, {stats.get('spans_count', 0)} spans\n")
    except Exception as e:
        print(f"❌ Cannot connect to AvocadoDB server: {e}")
        print("   Start server with: ./target/release/avocado-server")
        return
    
    # Example queries
    queries = [
        "How does the compile function work?",
        "What is the WorkingSet structure?",
        "How does span extraction work?",
    ]
    
    for query in queries:
        print(f"Question: {query}")
        print("-" * 60)
        
        try:
            # Use ask() method - automatically uses TinyLlama if available
            answer = db.ask(query, llm="auto")
            print(f"Answer: {answer}\n")
        except Exception as e:
            print(f"❌ Error: {e}\n")
            # Fallback to just context
            try:
                context = db.compile(query)
                print(f"Context (fallback): {context.text[:200]}...\n")
            except:
                pass
    
    print("=" * 60)
    print("\n💡 Tips:")
    print("   - Use llm='auto' (default) to try LLM, fallback to context")
    print("   - Use llm='local' to require TinyLlama")
    print("   - Use llm='none' to just get context (same as compile())")


if __name__ == "__main__":
    main()

